presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java

```diff
@@ -399,4 +401,149 @@ public long getRetainedSizeInBytes()
     {
         return INSTANCE_SIZE + sizeOf(isNullVector) + sizeOf(inDictionaryVector);
     }
+
+    @Override
+    public void scan()
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup(type);
+        }
+        beginScan(presentStream, null);
+        ensureValuesCapacity();
+        makeInnerQualifyingSet();
+
+        QualifyingSet input = hasNulls ? innerQualifyingSet : inputQualifyingSet;
+        if (filter != null) {
+            outputQualifyingSet.reset(input.getPositionCount());
+        }
+
+        if (input.getPositionCount() > 0) {
+            if (filter != null) {
+                int numInput = input.getPositionCount();
+                int[] innerToOuter = !hasNulls ? null : innerQualifyingSet.getInputNumbers();
+                int[] outerRows = !hasNulls ? input.getPositions() : inputQualifyingSet.getPositions();
+                QualifyingSet output = outputQualifyingSet;
+                output.reset(numInput);
+                numInnerResults = dataStream.scan(filter,
+                        input.getPositions(),
+                        0,
+                        numInput,
+                        input.getEnd(),
+                        outerRows,
+                        innerToOuter,
+                        output.getPositions(),
+                        output.getInputNumbers(),
+                        values,
+                        numValues);
+                output.setPositionCount(numInnerResults);
+            }
+            else {
+                numInnerResults = dataStream.scan(null,
+                        input.getPositions(),
+                        0,
+                        input.getPositionCount(),
+                        input.getEnd(),
+                        null,
+                        null,
+                        null,
+                        null,
+                        values,
+                        numValues);
+            }
+        }
+
+        if (inDictionaryStream != null) {
```

Comment:
Should there be a requireNonNull?
If inDictionaryStream can be null to indicate some other behavior can a comment be added to explain this?

Reply:
There are dictionaries at stripe and row-group levels. When all values refer to row-group dictionary, inDictionaryStream is null. If some rows refer to stripe-level dictionary, then inDictionaryStream is not null and has a bit set for these rows.